### ISSUE TITLE
`has_inherited_table` typing warning in PyCharm

### ISSUE THREAD

**Issue by caffeinatedMike** [sqlalchemy/sqlalchemy#8909]
### Describe the bug

Further typing refinement is required for `sqlalchemy.orm.has_inherited_table` as PyCharm currently complains when it's used both directly inside the `DeclarativeBase` class and in Mixin classes (as the [2.0 docs suggest](https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html#using-orm-declared-attr-with-inheriting-table-and-mapper-arguments)).

> Expected type 'Type[_O]', got 'Base' instead

### To Reproduce

```python
from typing import Optional
from sqlalchemy.orm import DeclarativeBase, declared_attr, has_inherited_table

# Scenario 1: Directly within a DeclarativeBase class
class Base(DeclarativeBase):
    @declared_attr.directive
    def __tablename__(cls) -> Optional[str]:
        if has_inherited_table(cls):  # <- PyCharm Warning: Expected type 'Type[_O]', got 'Base' instead
            return None
        return cls.__name__.lower()

# Scenario 2: Within a Mixin class (directly from the 2.0 docs)
class Tablename:
    @declared_attr.directive
    def __tablename__(cls):
        if has_inherited_table(cls):  # <- PyCharm Warning: Expected type 'Type[_O]', got 'Tablename' instead
            return None
        return cls.__name__.lower()
```

### Error

N/A - Non-runtime issue

### Versions

- OS: Microsoft Windows [Version 10.0.19043.2130]
- Python: 3.7.8 (tags/v3.7.8:4b47a5b6ba, Jun 28 2020, 08:53:46) [MSC v.1916 64 bit (AMD64)] on win32
- SQLAlchemy: 2.0.0b3
- Database: N/A, but PostgreSQL 14
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg

### Additional context

I prefer to have the logic inside the `Base(DeclarativeBase)` class, so the logic is applied conditionally to all mapped tables without needing a Mixin for this specific functionality.

**Comment by zzzeek** [comment]
The issue is with declared_attr, not has_inheited_table.  There's a workaround, however no fix without a workaround for this right now because it's not clear how to get typing tools to recognize them as treating a given function like a classmethod.

The workaround is just that.  Add `@classmethod`:

```py
from typing import Optional
from sqlalchemy.orm import DeclarativeBase, declared_attr, has_inherited_table

# Scenario 1: Directly within a DeclarativeBase class
class Base(DeclarativeBase):
    @declared_attr.directive
    @classmethod
    def __tablename__(cls) -> Optional[str]:
        if has_inherited_table(cls):  # <- PyCharm Warning: Expected type 'Type[_O]', got 'Base' instead
            return None
        return cls.__name__.lower()

# Scenario 2: Within a Mixin class (directly from the 2.0 docs)
class Tablename:
    @declared_attr.directive
    @classmethod
    def __tablename__(cls):
        if has_inherited_table(cls):  # <- PyCharm Warning: Expected type 'Type[_O]', got 'Tablename' instead
            return None
        return cls.__name__.lower()
```

**Comment by zzzeek** [comment]
> Good to know, thanks for the tip!
> 
> I know I'm probably way out of my depth here, but in the case of getting sqlalchemy's `declared_attr` to be recognized as a class method could the non-decorator usage of `classmethod` be employed? I'm thinking somewhere in the chain of the `declared_attr` setup the attribute instance wraps itself (ex: `classmethod(self)`). Not sure if that helps or is even possible, but figured I'd throw out some shower thoughts.

yes it might be something like that, though I might have tried that already.  the tools keep changing, but in the case of special things like `@property` and possibly `@classmethod`, they are often hardcoded and you can't use that construct in any other kind of way.

**Comment by caffeinatedMike** [comment]
True, I see what you mean. I know it's not a concrete solution, but in the case of `has_inherited_table` it seems to be resolved by updating the the signature to `def has_inherited_table(cls: Type[Union[_O, _OO]]) -> bool:`.

**Comment by zzzeek** [comment]
sure but that's not correct.   there's any number of functions one could invoke that accept a class as its argument which would all have this problem.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Thanks for the detailed report — the `### Error` error you hit makes the failure mode concrete. Before changing anything I'd like to reproduce it against current `main` first, because a fix anchored to observed behavior is much easier to review than one inferred from the report alone.

Plan: trace the path, reproduce it, then make the smallest change that addresses it, plus a regression test if this area has coverage. If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first? (relevant PRs, existing tests, known constraints)

(claiming via @LeonxLJX)
